flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NetworkEnvironmentConfiguration.java

```diff
@@ -105,6 +106,302 @@ public NettyConfig nettyConfig() {
+	 * @param maxJvmHeapMemory the maximum JVM heap size (in bytes)
+	 * @param localTaskManagerCommunication true, to skip initializing the network stack
+	 * @param taskManagerAddress identifying the IP address under which the TaskManager will be accessible
+	 * @return NetworkEnvironmentConfiguration
+	 */
+	@Deprecated
+	public static NetworkEnvironmentConfiguration fromConfiguration(
+		Configuration configuration,
+		long maxJvmHeapMemory,
+		boolean localTaskManagerCommunication,
+		InetAddress taskManagerAddress) {
+
+		// ----> hosts / ports for communication and data exchange
+
+		final int dataport = configuration.getInteger(TaskManagerOptions.DATA_PORT);
+		ConfigurationParserUtils.checkConfigParameter(dataport >= 0, dataport, TaskManagerOptions.DATA_PORT.key(),
+			"Leave config parameter empty or use 0 to let the system choose a port automatically.");
+
+		final int pageSize = ConfigurationParserUtils.getPageSize(configuration);
+
+		final int numNetworkBuffers;
+		if (!hasNewNetworkConfig(configuration)) {
+			// fallback: number of network buffers
+			numNetworkBuffers = configuration.getInteger(TaskManagerOptions.NETWORK_NUM_BUFFERS);
+
+			checkOldNetworkConfig(numNetworkBuffers);
+		} else {
+			if (configuration.contains(TaskManagerOptions.NETWORK_NUM_BUFFERS)) {
+				LOG.info("Ignoring old (but still present) network buffer configuration via {}.",
+					TaskManagerOptions.NETWORK_NUM_BUFFERS.key());
+			}
+
+			final long networkMemorySize = calculateNewNetworkBufferMemory(configuration, maxJvmHeapMemory);
+
+			// tolerate offcuts between intended and allocated memory due to segmentation (will be available to the user-space memory)
+			long numNetworkBuffersLong = networkMemorySize / pageSize;
+			if (numNetworkBuffersLong > Integer.MAX_VALUE) {
+				throw new IllegalArgumentException("The given number of memory bytes (" + networkMemorySize
+					+ ") corresponds to more than MAX_INT pages.");
+			}
+			numNetworkBuffers = (int) numNetworkBuffersLong;
+		}
+
+		final NettyConfig nettyConfig;
+		if (!localTaskManagerCommunication) {
+			final InetSocketAddress taskManagerInetSocketAddress = new InetSocketAddress(taskManagerAddress, dataport);
+
+			nettyConfig = new NettyConfig(taskManagerInetSocketAddress.getAddress(), taskManagerInetSocketAddress.getPort(),
+				pageSize, ConfigurationParserUtils.getSlot(configuration), configuration);
+		} else {
+			nettyConfig = null;
+		}
+
+		int initialRequestBackoff = configuration.getInteger(TaskManagerOptions.NETWORK_REQUEST_BACKOFF_INITIAL);
+		int maxRequestBackoff = configuration.getInteger(TaskManagerOptions.NETWORK_REQUEST_BACKOFF_MAX);
+
+		int buffersPerChannel = configuration.getInteger(TaskManagerOptions.NETWORK_BUFFERS_PER_CHANNEL);
+		int extraBuffersPerGate = configuration.getInteger(TaskManagerOptions.NETWORK_EXTRA_BUFFERS_PER_GATE);
+
+		boolean isCreditBased = configuration.getBoolean(TaskManagerOptions.NETWORK_CREDIT_MODEL);
```

Comment:
It is actually for the consideration of tests. 
We have the `NetworkConfiguraitonBuilder` which is easy for setting `isCreditBased` variable in tests, otherwise it needs to construct the `NettyConfig` with internal `Configuration` to set this variable. I ever considered removing the corresponding `getter` from `NettyConfig` to avoid duplication.

Reply:
As I understand `NetworkEnvironmentConfigurationBuilder.nettyConfig` is `null` by default.
Then maybe `nettyConfig` could be some created default non-null object and `NetworkEnvironmentConfigurationBuilder` could also have `setDefaultNettyConfigAndIsCreditBased(boolean isCreditBased)` instead of `setIsCreditBased`? Internally it would construct default non-null `nettyConfig` with requested `isCreditBased`.

Otherwise, it might be confusing for future code maintaining to have two sources of `isCreditBased`.

Reply:
I also do not like the way of two sources of `isCreditBased`. 

I think the key problem is whether `isCreditBased` must exist together with `NettyConfig`. My previous thought is trying to make decouple `isCreditBased` with `NettyConfig`. 

We know `NettyConfig` is optional currently in `NetworkEnvironmentConfiguration`, and most of the existing tests should verify the default behavior (`isCreditBased = true`) and only several tests would verify both modes. So `NetworkEnvironmentConfigurationBuilder` would create `NetworkEnvironmentConfiguration` with `isCreditBased = true` and null `NettyConfig` by default which is suitable and simple for most of the tests.

If we remove `isCreditBased` from `NetworkEnvironmentConfiguration`, the `NetworkEnvironmentConfigurationBuilder` must create a default `NettyConfig` with enabling credit-based by default.

So it might be better if we make `isCreditBased` property independent with `NettyConfig`. Another option is we remove `NettyConfig#isCreditBasedEnabled` because via `NettyConfig#getConfig` we could get everything including the `isCreditBased`, not need that explicit method for every property. To do so it might seem no duplications between `NettyConfig` and `NetworkEnvironmentConfiguration`. What do you think?

Reply:
In general, I do not see a problem to set default `NettyConfig` with enabling credit-based by default in `NetworkEnvironmentConfigurationBuilder`. It is just some constant `NettyConfig` or what is problem here? It also looks like inconsistent `NetworkEnvironmentConfiguration` with `isCreditBased = true` and `NettyConfig = null` (so there is somewhat coupling). I would say `NetworkEnvironmentConfigurationBuilder.setNettyConfig(null)` should also set `isCreditBased = false`.

`NetworkEnvironmentConfiguration.isCreditBased` could be:
`return nettyConfig != null && nettyConfig.isCreditBasedEnabled();`

Reply:
Yes, it actually has no problem to set default `NettyConfig` with enabling credit-based by default in `NetworkEnvironmentConfigurationBuilder`.

My previous thought is that even though we do not use netty technology in network stack, the credit-based mode might also suitable for other network frameworks. In other words, the credit-based is more related to `ResultPartition` and `SingleInputGate`, no matter with the transport layer is netty or not. So if `NettyConfig` is null, the credit-based property could also be used in other parts.

Certainly I could approve the way of creating default `NettyConfig` instead of current null in `NetworkEnvironmentConfigurationBuilder`. :)

Reply:
ok, sorry, I see your point now. Actually, the credit-based mode seems to belong to Network but not Netty. Maybe, `NettyConnectionManager` should rather take one more constructor argument `isCreditBased` and then we can keep it only in `NetworkEnvironmentConfiguration`. We should just make sure that `isCreditBased` is `false` for `NetworkEnvironmentConfiguration.nettyConfig = null`

Reply:
I even considered the way of adding one more parameter `isCreditBased` in `NettyConnectionManager`. But `TaskManagerOptions.NETWORK_CREDIT_MODEL` is `Deprecated`, and it is supposed to be removed from FLINK-1.6. I think this temporary option would be abandoned finally after one undergoing network issue confirmed. So it might be not good to reflect it in the constructor of `NettyConnectionManager`  for avoiding changing repeatedly.

`NettyConfig` has not maintained the `isCreditBased` property explicitly, and it just gets this option from internal `Configuration` via `isCreditBasedEnabled` method. So we could remove this method then `NettyConnectionManager` could still get it via `client.getConfig().getBoolean(TaskManagerOptions.NETWORK_CREDIT_MODEL)`. I think this property seems not related to `NettyConfig` in this way.

In addition, the current netty stack could support both credit and non-credit modes in two suits of codes. So if `isCreditBased` is false, that does not mean `NettyConfig` is null. `NettyConfig` is null or not is based on whether downstream and upstream are in the same task manager.

Reply:
After thinking through this issue again, if we keep maintaining `isCreditBased` in `NetworkEnvironmentConfiguration`, we have to confirm this property consistent between `NetworkEnvironmentConfiguration` and Flink `Configuration` in `NettyConfig`. So the codes seem ugly in `NetworkEnvironmentConfigurationBuilder`. Then atm I would prefer to add this property in the constructor of `NettyConnectionManager` even though we would remove this deprecated option future.

I have submitted the fixup commit for this issue. :)